### PR TITLE
Feature: OAuth2 Gmail and Outlook email support

### PR DESCRIPTION
**Pull request by shamoon**
## Proposed change

In the end I think this wasn't so bad, once I figured out the respective flows etc.

Welcome any feedback in general, of course, and specifically if anyone has good understanding of oauth etc in case I've missed anything here

Screenshot 2024-10-06 at 2 42 30 PM

https://github.com/user-attachments/assets/18868502-03ea-435c-8c3a-aa3b63487936

Closes #4063

## Type of change

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by stumpylog** [comment]
I have nothing constructive as of yet, but did you have a look at https://frankie567.github.io/httpx-oauth/?  Maybe it could simplify things for you?

**Comment by shamoon** [comment]
> I have nothing constructive as of yet, but did you have a look at https://frankie567.github.io/httpx-oauth/? Maybe it could simplify things for you?

Ha, wish I had seen that before. I'd be happy to try and integrate it, it would cut out some of my code here, only downside I suppose would be the extra dependency?

**Comment by shamoon** [comment]
Lol, diff is almost the same n lines, but still I agree better to use the library =)

File: docs/configuration.md [hunk 1]
```diff
@@ -1213,6 +1207,48 @@ consumers working on the same file. Configure this to prevent that.
...
+
+### Email OAuth {#email_oauth}
+
+#### [`PAPERLESS_OAUTH_CALLBACK_BASE_URL=<str>`](#PAPERLESS_OAUTH_CALLBACK_BASE_URL) {#PAPERLESS_OAUTH_CALLBACK_BASE_URL}
+
+: The base URL for the OAuth callback. This is used to construct the full URL for the OAuth callback. This should be the URL that the Paperless instance is accessible at. If not set, defaults to the `PAPERLESS_URL` setting. At least one of these settings must be set to enable OAuth Email setup.
+
+    Defaults to none (thus will use [PAPERLESS_URL](#PAPERLESS_URL)).
+
+#### [`PAPERLESS_GMAIL_OAUTH_CLIENT_ID=<str>`](#PAPERLESS_GMAIL_OAUTH_CLIENT_ID) {#PAPERLESS_GMAIL_OAUTH_CLIENT_ID}
```
Comment (stumpylog) [inline docs/configuration.md:1220]: So with the client id and secret set here, does that only allow a single gmail and single microsoft account?

File: src/paperless_mail/oauth.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    )
+
+    def __init__(self):
+        self._gmail_client = None
+        self._outlook_client = None
+
+    @property
+    def gmail_client(self) -> OAuth2:
+        if self._gmail_client is None:
+            self._gmail_client = OAuth2(
```
Comment (stumpylog) [inline src/paperless_mail/oauth.py:31]: I know, still not helpful, but was there a reason to not use [GoogleOAuth2 ](https://frankie567.github.io/httpx-oauth/reference/httpx_oauth.clients/#httpx_oauth.clients.google.GoogleOAuth2) or [MicrosoftGraphOAuth2](https://frankie567.github.io/httpx-oauth/reference/httpx_oauth.clients/#httpx_oauth.clients.microsoft.MicrosoftGraphOAuth2)?

I promise I'll actually try this soon (tm)

File: docs/configuration.md (hunk 1 above)
Comment (shamoon) [reply docs/configuration.md:1220]: Nope, you can re-use these for as many accounts as you like. For Microsoft I think the settings (in Azure) can be relevant, e.g. you can setup your 'app' to only allow users in your domain etc but that would just cause auth to fail if you used e.g. an @outlook.com account

File: src/paperless_mail/oauth.py (hunk 2 above)
Comment (shamoon) [reply src/paperless_mail/oauth.py:31]: Haha, no rush. I tried them at first and they didnt work, but I also fixed some other stuff after that which may have been the cause. I can circle back and try them again.

**Review by stumpylog — APPROVED** [review]

**Comment by RincewindX** [comment]
For me it's not shown in the frontend. Am i missing something?
grafik

**Comment by agrmohit** [comment]
You need to create your own OAuth apps and set it in configuration in order to use it. You can find instructions here: https://docs.paperless-ngx.com/usage/#oauth-email-setup

Paperless-ngx doesn't ship with any default configs for them.

**Pull request merged**